Add a new "is-task" ps filter

vdemeester opened the pull request:
This is a _simple_ try at #24393 🐳. 

This PR add a `is-task` filter to `docker ps` to make it easier to list containers that are managed (started with swarm mode) and those who are not 👼.

The accepted values of this new `is-task` filters are : `true`, `false`.

``` bash
$ docker service create --replicas 3 --name toto busybox top
$ docker run -itd busybox sh
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
6b285d27cb6b        busybox             "sh"                14 minutes ago      Up 14 minutes                           angry_albattani
0386186609e5        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.3.ecv0v800vjm5tczzqjjez4q9w
4d5f0bb784d0        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.2.7s1mw00ssx5pl4i938kya3zsi
d107a6069a97        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.1.46ppfac4ol5mlrh4q6cpsuy88
$ docker ps --filter=is-task=true
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
0386186609e5        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.3.ecv0v800vjm5tczzqjjez4q9w
4d5f0bb784d0        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.2.7s1mw00ssx5pl4i938kya3zsi
d107a6069a97        busybox:latest      "top"               14 minutes ago      Up 14 minutes                           toto.1.46ppfac4ol5mlrh4q6cpsuy88
$ docker ps --filter=is-task=true
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
6b285d27cb6b        busybox             "sh"                14 minutes ago      Up 14 minutes                           angry_albattani
$ docker ps --filter=is-task=wrong
Error response from daemon: Invalid filter 'is-task=[wrong]'
```

<del>As @thaJeztah said, "I'm completely open to suggestions w.r.t. naming; is "managed" the right term for this?" 👼</del>


If this pass design-review, I'll need to:
- [x] Add some tests
- [X] Update completion
- [x] Update documentation

Closes #24393 

/cc @thaJeztah @icecrime @stevvooe @aaronlehmann @dnephin 

Signed-off-by: Vincent Demeester vincent@sbr.pm

albers commented:
Do we need `any` here? IIUC, filtering with a value of `any` is just the same as omitting the filter.
The `dangling` filters, which are very similar, do not have a value of `any`.

In contrast, I like the `any` setting. It improves the UX. How about keeping it here and adding `any` to the existing boolean filters as well?

justincormack commented:
I think I slightly prefer that there is no `any` value, but not strongly, but I do think consistency is important so maybe remove it now and consider a separate PR to add any everywhere?

thaJeztah commented:
thinking of this; if we keep it as a boolean, we could use `is-managed` as a name

vdemeester commented:
@albers @thaJeztah @justincormack good points :stuck_out_tongue_closed_eyes: 
Just waiting a little longer to have more opinion on this, but I would be ok to make it a boolean :angel:

stevvooe commented:
This can already be done with `docker ps -f label=com.docker.swarm.task`. I could see a need for having this alias `docker ps -f task=true` but the term "managed" isn't quite right.

thaJeztah commented:
Alright, let's use "task"; since it's a boolean, I'm suggesting to use `is-task` for consistency (and keeping `task` open for future expansion to filter on a specific task)

@stevvooe SGTY?

stevvooe commented:
SGTM

icecrime commented:
We're getting closer to 1.12.0 release date, and I'd like to keep the absolute minimum on the milestone. We can decide to add this anytime (including a patch release), so I'll remove this one from the milestone (sorry!).

vdemeester commented:
@icecrime no problemo, make sense :wink:

vdemeester commented on docs/reference/commandline/ps.md:
```diff
@@ -27,6 +27,7 @@ Options:
                         - since=(<container-name>|<container-id>)
                         - ancestor=(<image-name>[:tag]|<image-id>|<image@digest>)
                           containers created from an image or a descendant.
+                        - managed=(true|false|any)
```
oh… I forgot to update that :sweat:

thaJeztah commented:
Can you update the PR / docs? then we can get this moving again :heart:

vdemeester commented:
@thaJeztah updated :angel:

cpuguy83 commented:
Can you add a test?

vdemeester commented:
I'll look into it 👼

vdemeester commented:
@cpuguy83 updated 👼

cpuguy83 commented on integration-cli/docker_cli_swarm_test.go:
```diff
@@ -242,3 +242,33 @@ func (s *DockerSwarmSuite) TestSwarmServiceWithGroup(c *check.C) {
+	c.Assert(strings.TrimSpace(out), checker.Not(checker.Equals), "")
+
+	// make sure task has been deployed.
+	waitAndAssert(c, defaultReconciliationTimeout, d.checkActiveContainerCount, checker.Equals, 1)
```
Shouldn't this be 2?

cpuguy83 replied:
btw, there's also `checkServiceRunningTasks`

vdemeester replied:
ah you're right, I'll update :wink:

stevvooe commented on docs/reference/commandline/ps.md:
```diff
@@ -27,6 +27,7 @@ Options:
                         - since=(<container-name>|<container-id>)
                         - ancestor=(<image-name>[:tag]|<image-id>|<image@digest>)
                           containers created from an image or a descendant.
+                        - is-task=(true|false|any)
```
What other areas do we prefix booleans with "is"? This is extremely awkward looking.

thaJeztah replied:
@stevvooe We use it for `docker search` (is-official, is-automated); https://docs.docker.com/engine/reference/commandline/search/#/filtering, possibly other locations that I cannot recall right now :smile:

vdemeester replied:
(oh I forgot to remove `any` from here)

stevvooe replied:
@thaJeztah Those are pretty archaic APIs. What is wrong with just `task`?

thaJeztah replied:
just by looking at "is-task", a user can deduct it's a boolean. I don't think just "task" is very descriptive, consider:

```
--filter task=true
--filter ancestor=true
```

First checks for a boolean value, second for an ancestor image named "true"

stevvooe replied:
Could it just be `--filter task`? We already have `ancestor` and not `is-ancestor`.

vdemeester replied:
@stevvooe make sence, I'll update to `task`

thaJeztah replied:
But we don't have `is-ancestor` because it's not a Boolean

vdemeester replied:
@thaJeztah ah you're right :sweat_smile:

thaJeztah commented:
@vdemeester does the current implementation actually support the "any" value? Reading back this thread, I think I'd prefer to not implement "any" as part of this PR, and (if we want to add it) do a separate PR that adds support for "any" to all booleans.

vdemeester commented:
@thaJeztah implementation does not support `any` as value (removed it), that's why I commented out that I forgot to update a doc file :stuck_out_tongue_closed_eyes:

thaJeztah commented:
LGTM, but needs a rebase and an update to the docs 🐰

vdemeester commented:
@thaJeztah done 👼

thaJeztah commented:
@vdemeester oh boy, needs another rebase :cry:

thaJeztah commented:
ping @justincormack  PTAL

vdemeester commented:
Rebased 👼

dnephin commented:
LGTM

thaJeztah commented:
moving to docs review

thaJeztah commented:
@vdemeester can you update the man page as well? https://github.com/docker/docker/blob/master/man/docker-ps.1.md

Also, the API docs mention filters here; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.25.md#list-containers

And should be in the API changes;
https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes

vdemeester committed:
Add a new "is-task" ps filter

This makes it easier to list containers that are part of a task
(swarm mode) and those who are not.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
@thaJeztah updated :angel:

thaJeztah reviewed (approved):
LGTM

vdemeester commented:
WindowsRS1 failure unrelated, merging 👼
